pkg/clusteragent/autoscaling/externalmetrics/errors.go

```diff
@@ -8,119 +8,101 @@
-func NewInvalidMetricNotFoundError(query string) *InvalidMetricNotFoundError {
-	return &InvalidMetricNotFoundError{
-		Query: query,
+func (e *queryError) Error() string {
+	if !e.retryAfter.IsZero() {
+		return fmt.Sprintf("Processing data from API failed, reason: %s, query was: %s, will retry at: %s", e.reason, e.query, e.retryAfter.Format(time.DateTime))
 	}
+	return fmt.Sprintf("Processing data from API failed, reason: %s, query was: %s", e.reason, e.query)
 }
 
-// InvalidMetricGlobalError represents a global backend error for all queries.
-type InvalidMetricGlobalError struct{}
-
-func (e *InvalidMetricGlobalError) Error() string {
-	return "Global error (all queries) from backend, invalid syntax in query? Check Cluster Agent leader logs for details"
-}
+func (e *queryError) Is(err error) bool {
+	queryErr, ok := err.(*queryError)
+	if !ok {
+		return false
+	}
 
-// NewInvalidMetricGlobalError initializes an InvalidMetricGlobalError.
-func NewInvalidMetricGlobalError() *InvalidMetricGlobalError {
-	return &InvalidMetricGlobalError{}
+	// We skip comparison of exact time, it's enough to be both sets or both zero.
+	return queryErr.query == e.query &&
+		queryErr.reason == e.reason &&
+		((queryErr.retryAfter.IsZero() && e.retryAfter.IsZero()) || (!queryErr.retryAfter.IsZero() && !e.retryAfter.IsZero()))
 }
 
-// InvalidMetricGlobalErrorWithRetries represents a global backend error for all queries with retry information.
-type InvalidMetricGlobalErrorWithRetries struct {
-	BatchSize  int
-	RetryAfter string
+// batchError represents a global backend error for all queries.
+type batchError struct {
+	err        error
+	retryAfter time.Time
 }
 
-func (e *InvalidMetricGlobalErrorWithRetries) Error() string {
-	return fmt.Sprintf("Global error (all queries, batch size %d) from backend, invalid syntax in query? Check Cluster Agent leader logs for details. Will retry after %s", e.BatchSize, e.RetryAfter)
+func newBatchError(err error, retryAfter time.Time) *batchError {
+	return &batchError{
+		err:        err,
+		retryAfter: retryAfter,
+	}
 }
 
-// NewInvalidMetricGlobalErrorWithRetries initializes an InvalidMetricGlobalErrorWithRetries.
-func NewInvalidMetricGlobalErrorWithRetries(batchSize int, retryAfter string) *InvalidMetricGlobalErrorWithRetries {
-	return &InvalidMetricGlobalErrorWithRetries{
-		BatchSize:  batchSize,
-		RetryAfter: retryAfter,
+func (e *batchError) Error() string {
+	if !e.retryAfter.IsZero() {
+		return fmt.Sprintf("Datadog API call error (all queries), error: %s, will retry at: %s", e.err, e.retryAfter.Format(time.DateTime))
 	}
+	return fmt.Sprintf("Datadog API call error (all queries), error: %s", e.err)
 }
 
-// RateLimitError represents a rate limit error.
-type RateLimitError struct {
-	Err error
+func (e *batchError) Unwrap() error {
+	return e.err
 }
 
-func (e *RateLimitError) Error() string {
-	return "Global error: Datadog API rate limit exceeded"
-}
+func (e *batchError) Is(err error) bool {
```

Comment:
You have several public function in this file. I suppose the linter will complains that you haven't documented them

Reply:
The linter ran ok so I guess it understands it's methods implementing the standard `error` interface